builtin/v9/migration/top.go

```diff
@@ -244,10 +244,12 @@ func MigrateStateTree(ctx context.Context, store cbor.IpldStore, newManifestCID
 		return cid.Undef, xerrors.Errorf("failed to find datacap code ID: %w", err)
 	}
 
-	if err = actorsOut.SetActor(builtin.DatacapActorAddr, &Actor{
+	if err = actorsIn.SetActor(builtin.DatacapActorAddr, &Actor{
 		Code: dataCapCode,
-		// we just need to put _something_ defined, this never gets read
-		Head:       emptyMapCid,
+		// we supply the verified registry's head here for caching purposes
+		// if the verifreg head has changed, a fresh migration is needed
+		// if not, the cached datacap migration can be used
+		Head:       verifregActorV8.Head,
```

Comment:
I don't understand how the regular [cached migration](https://github.com/filecoin-project/go-state-types/pull/89/files#diff-0f71e4e5376421454dd6e1792714840e621766fb31e279de1749157bf59a8bd3R554) is going to map this to the correct thing in the case of a cache hit.  Why isn't it mapping cached verifreg in to verifreg out?  Where is the logic to swap the verifreg state for datacap state?

Reply:
Despite whatever I'm missing that makes the above work this approach is too clever for its own good.  cachedMigration does one simple thing which is distinct from what we want here.  I think you want to define a new migration transform like cachedMigration that explicitly checks if verifreg is in the cache and then migrates datacap.

Reply:
TL;DR: I agree with you, we can just drop the caching here -- it's a little sad because we will very likely have a cache hit (the verifreg changes infrequently), but it's likely not the critical path and not worth a bug.

The caching, as I understand it, is keyed by an `Address, Head` tuple. During the premigration, we map `Address, Head` to the `MigratedHead`. If at the time of the actual migration we have a hit for `Address, Head`, we blindly use `MigratedHead`.

The DataCap actor is in the peculiar position of being loadable from the cache if the verifreg head doesn't change. Thus, if we enter `DataCapAddr, VerifregHead` into our cache, mapping it to `NewDataCapHead`, we'll correctly load it. 

Note that we will also use the cached entry for the Verifreg migration, which will have been entered with a key of `VerifregAddr, VerifregHead`. 

But, all in all, this is janky. Let's see if we can make do with no caching for the DataCap actor?

Reply:
I think you should make use of this caching I just want it done in a way that removes faking out the datacap actor's head cid.  If you implement a variation on `cachedMigration` then this would be nice.  I.e. define `conditionalMigrator` that tracks two addresses and checks the cache for the first one before checking the cache for the second one.